### PR TITLE
Feature/tecg 296 follow technology

### PR DESCRIPTION
**Pull request by felipeaziz**

### CODE REVIEW

File: src/main/java/com/ciandt/techgallery/persistence/model/TechnologyFollowers.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  /*
+   * Constants --------------------------------------------
+   */
+  public static final String ID = "Id";
+  public static final String TECHNOLOGY = "technology";
+  public static final String FOLLOWERS = "followers";
+
+  @Id
+  Long id;
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/persistence/model/TechnologyFollowers.java]: Please use the same id as Technology. Use a String id and fill it with the technology id.
When using transactions is preferred to have a known id.

File: src/main/java/com/ciandt/techgallery/service/impl/TechnologyServiceImpl.java [hunk 2]
```diff
@@ -397,4 +405,38 @@ public void updateEdorsedsCounter(Technology technology, Integer size) {
     technology.setEndorsersCounter(size);
     technologyDAO.update(technology);
   }
+
+  @Override
+  public Technology followTechnology(Technology technology, User user)
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/impl/TechnologyServiceImpl.java]: Esse método deveria estar na TechnologyFollowersService. Também quebre em mais métodos como deleteTechnologyFollowers, addTechnologyFollowers

File: src/main/java/com/ciandt/techgallery/persistence/dao/impl/TechnologyCommentDAOImpl.java [hunk 3]
```diff
@@ -52,4 +53,17 @@ public static TechnologyCommentDAOImpl getInstance() {
...
     return entities;
   }
+
+  @Override
+  public String findAllCommentsIdsStartingFrom(Date date) {
+    final String commentsIds = "";
+    final Objectify objectify = OfyService.ofy();
+    final List<TechnologyComment> comments =
+        objectify.load().type(TechnologyComment.class)
+            .filter(TechnologyComment.TIMESTAMP + " >" , date).list();
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/persistence/dao/impl/TechnologyCommentDAOImpl.java]: Should not ">=" ?

File: src/main/java/com/ciandt/techgallery/persistence/dao/impl/TechnologyRecommendationDAOImpl.java [hunk 4]
```diff
@@ -84,4 +85,19 @@ public TechnologyRecommendation findByComment(TechnologyComment comment) {
     }
   }
 
+  @Override
+  public String findAllRecommendationsIdsStartingFrom(Date date) {
+    final Objectify objectify = OfyService.ofy();
+    final List<TechnologyRecommendation> allRecommendations =
+        objectify.load().type(TechnologyRecommendation.class).list();
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/persistence/dao/impl/TechnologyRecommendationDAOImpl.java]: @felipeaziz  Load all Recommendations is not a good practice.

File: src/main/java/com/ciandt/techgallery/persistence/model/EmailNotification.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.util.List;
+
+@Entity
+public class EmailNotification extends BaseEntity<Long> {
+
+  @Id
+  private Long id;
+  @Index
+  private String technologyId;
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/persistence/model/EmailNotification.java]: @brunoliberal Why do you need the technologyid?

File: src/main/java/com/ciandt/techgallery/persistence/model/TechnologyFollowers.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  /*
+   * Constants --------------------------------------------
+   */
+  public static final String ID = "id";
+  public static final String TECHNOLOGY = "technology";
+  public static final String FOLLOWERS = "followers";
+
+  @Id
+  String id;
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/persistence/model/TechnologyFollowers.java]: @felipeaziz Please document the id field as it is the same id from Technology.

File: src/main/java/com/ciandt/techgallery/persistence/dao/impl/TechnologyRecommendationDAOImpl.java (hunk 4 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/persistence/dao/impl/TechnologyRecommendationDAOImpl.java]: The problem here is that to obtain the timestamp from a recommendation we have to get its comment and get its timestamp then compare it. Maybe if we had another timestamp inside recommendation is a better practice.

File: src/main/java/com/ciandt/techgallery/service/impl/EmailServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    }
+    return from;
+  }
+
+  /**
+   * Push one email to email queue for each follower user in each technology that have
+   * followers.
+   */
+  public void sendEmailtoFollowers() {
+    CronJob cronJob = new CronJob();
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/impl/EmailServiceImpl.java]: @brunoliberal EmailService should only send emails. Create a service class for this cron job.

File: src/main/java/com/ciandt/techgallery/persistence/model/TechnologyFollowers.java (hunk 6 above)
Comment (felipeaziz) [inline src/main/java/com/ciandt/techgallery/persistence/model/TechnologyFollowers.java]: Done

File: src/main/java/com/ciandt/techgallery/persistence/dao/impl/TechnologyRecommendationDAOImpl.java (hunk 4 above)
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/persistence/dao/impl/TechnologyRecommendationDAOImpl.java]: As we talked, we defined that TechnologyRecommendation must have a timestamp as well.

File: src/main/java/com/ciandt/techgallery/persistence/model/EmailNotification.java (hunk 5 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/persistence/model/EmailNotification.java]: My mistake. its removed now.

File: src/main/java/com/ciandt/techgallery/service/impl/EmailServiceImpl.java (hunk 7 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/service/impl/EmailServiceImpl.java]: Created

File: src/main/java/com/ciandt/techgallery/service/email/EmailConfig.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+ */
+public class EmailConfig {
+
+  private String[] to;
+  
+  private String subject;
+
+  private String body;
+  
+  private String transferId;
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/email/EmailConfig.java]: @brunoliberal  This does not belong to TechGallery

File: src/main/java/com/ciandt/techgallery/service/impl/TechnologyFollowersServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+import java.util.ArrayList;
+
+/**
+ * Services for Technology Endpoint requests.
+ *
+ * @author felipers
+ *
+ */
+public class TechnologyFollowersServiceImpl implements TechnologyFollowersService {
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/impl/TechnologyFollowersServiceImpl.java]: @felipeaziz :  I forgot to tell you that you must add @Transactional annotation on the class declaration as well.

Comment (felipeaziz) [inline src/main/java/com/ciandt/techgallery/service/impl/TechnologyFollowersServiceImpl.java]: Done

File: src/main/java/com/ciandt/techgallery/service/email/EmailConfig.java (hunk 8 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/service/email/EmailConfig.java]: My mistake. 
Its fixed now.

**Pull request merged**